2.9BSD/usr/man/cat1/ls.1

LS(1)               UNIX Programmer's Manual                LS(1)

NAME
     ls - list contents of directory

SYNOPSIS
     ls [ -abcdfgilmnqrstux1ACFR ] name ...
     l [ _l_s options ] name ...

DESCRIPTION
     For each directory argument, _l_s lists the contents of the
     directory; for each file argument, _l_s repeats its name and
     any other information requested.  The output is sorted
     alphabetically by default.  When no argument is given, the
     current directory is listed.  When several arguments are
     given, the arguments are first sorted appropriately, but
     file arguments appear before directories and their contents.

     There are three possible listing formats.  The format chosen
     depends on whether the output is going to a teletype, and
     may also be controlled by option flags.  The default format
     for a teletype is to list the contents of directories in
     multi-column format, with the entries sorted down the
     columns.  (See also the -F and -x options below.) If the
     standard output is not a teletype, the default format is to
     list one entry per line.  Finally, there is a stream output
     format in which files are listed across the page, separated
     by `,' characters.  The -m flag enables this format.

     When invoked as _l the default is -m .

     Note: Files which are not the contents of a directory being
     interpreted are always sorted across the page rather than
     down the page in columns.  This is because the individual
     file names may be arbitrarily long.

     There are several thousand options:

     -l   List in long format, giving mode, number of links,
          owner, size in bytes, and time of last modification for
          each file.  (See below.) If the file is a special file
          the size field will instead contain the major and minor
          device numbers.  If the file is a quota node, the
          current usage, quota, and percentage of quota in use
          are printed.

     -t   Sort by time modified (latest first) instead of by
          name, as is normal.

     -a   List all entries; usually filenames beginning with `.'
          are suppressed.

     -A   List `.' entries; but suppress `.' and `..'.

Printed 7/25/83                                                 1

LS(1)               UNIX Programmer's Manual                LS(1)

     -s   Give size in blocks (1024 bytes), including indirect
          blocks, for each entry.

     -d   If argument is a directory, list only its name, not its
          contents (mostly used with -l to get status on direc-
          tory).

     -r   Reverse the order of sort to get reverse alphabetic or
          oldest first as appropriate.

     -u   Use time of last access instead of last modification
          for sorting (-t) or printing (-l).

     -c   Use time of file creation for sorting or printing.

     -i   Print i-number in first column of the report for each
          file listed.

     -f   Force each argument to be interpreted as a directory
          and list the name found in each slot.  This option
          turns off -l, -t, -s, and -r, and turns on -a; the
          order is the order in which entries appear in the
          directory.

     -g   Give group instead of owner in long listing.

     -n   With -l or -g gives group-ID or owner-ID instead of
          owner or group name.

     -m   force stream output format

     -1   force one entry per line output format, e.g. to a tele-
          type

     -C   force multi-column output, e.g. to a file or a pipe

     -q   force printing of non-graphic characters in file names
          as the character `?'; this normally happens only if the
          output device is a teletype

     -b   force printing of non-graphic characters to be in the
          \ddd notation, in octal.

     -x   force columnar printing to be sorted across rather than
          down the page; this is the default if the last charac-
          ter of the name the program is invoked with is an `x'.

     -F   cause directories to be marked with a trailing `/' and
          executable files to be marked with a trailing `*'; this
          is the default if the last character of the name the
          program is invoked with is a `f'.

Printed 7/25/83                                                 2

LS(1)               UNIX Programmer's Manual                LS(1)

     -R   recursively list subdirectories encountered.

     The mode printed under the -l option contains 11 characters
     which are interpreted as follows: the first character is

     b  if the entry is a block-type special file;
     c  if the entry is a character-type special file;
     d  if the entry is a directory;
     l  if the entry is a symbolic link;
     q  if the entry is a quota;
     -  if the entry is a plain file.

     The next 9 characters are interpreted as three sets of three
     bits each.  The first set refers to owner permissions; the
     next to permissions to others in the same user-group; and
     the last to all others.  Within each set the three charac-
     ters indicate permission respectively to read, to write, or
     to execute the file as a program.  For a directory, `exe-
     cute' permission is interpreted to mean permission to search
     the directory for a specified file.  The permissions are
     indicated as follows:

     r  if the file is readable;
     w  if the file is writable;
     x  if the file is executable;
     -  if the indicated permission is not granted.

     The group-execute permission character is given as s if the
     file has set-group-ID mode; likewise the user-execute per-
     mission character is given as s if the file has set-user-ID
     mode.

     The last character of the mode (normally `x' or `-') is t if
     the 1000 bit of the mode is on.  See _c_h_m_o_d(1) for the mean-
     ing of this mode.

     When the sizes of the files in a directory are listed, a
     total count of blocks, including indirect blocks is printed.

FILES
     /etc/passwd         to get user ID's for `ls -l'.
     /etc/group          to get group ID's for `ls -g'.

SEE ALSO
     file(1), pq(1), stat(2)

BUGS
     Newline and tab are considered printing characters in file
     names.

     The output device is assumed to be 80 columns wide.  Columns
     are basically 15 characters wide, 20 characters wide if -i

Printed 7/25/83                                                 3

LS(1)               UNIX Programmer's Manual                LS(1)

     or -s is specified, 25 characters wide if both -i and -s are
     specified.  In any case where -l or -n is specified, we
     assume only one entry will fit per line.

     The option setting based on whether the output is a teletype
     is undesirable as:

          ls -s | lpr

     is much different from

          ls -s

     On the other hand, not doing this setting would make old
     shell scripts which used _l_s almost certain losers, and then
     one could no longer do, e.g.:

          ls -s | sort -nr

     but would have to do

          ls -s -1 | sort -nr

     Column widths choices are poor for terminals which can tab.

Printed 7/25/83                                                 4



2.9BSD/usr/src/cmd/ls/ls.c
/*
 * list file or directory
 */

#include <sys/param.h>
#include <sys/stat.h>
#include <sys/dir.h>
#include <stdio.h>

#define	NFILES	1024
FILE	*pwdf, *dirf;
char	stdbuf[BUFSIZ];

struct lbuf {
	union {
		char	lname[15];
		char	*namep;
	} ln;
	char	ltype;
	short	lnum;
	short	lflags;
	short	lnl;
	short	luid;
	short	lgid;
	long	lsize;
	long	lmtime;
};

int	aflg, dflg, lflg, sflg, tflg, uflg, iflg, fflg, gflg, cflg;
int	rflg	= 1;
long	year;
int	flags;
int	lastuid	= -1;
char	tbuf[16];
long	tblocks;
int	statreq;
struct	lbuf	*flist[NFILES];
struct	lbuf	**lastp = flist;
struct	lbuf	**firstp = flist;
char	*dotp	= ".";

char	*makename();
struct	lbuf *gstat();
char	*ctime();
long	nblock();

#define	ISARG	0100000

main(argc, argv)
char *argv[];
{
	int i;
	register struct lbuf *ep, **ep1;
	register struct lbuf **slastp;
	struct lbuf **epp;
	struct lbuf lb;
	char *t;
	int compar();

	setbuf(stdout, stdbuf);
	time(&lb.lmtime);
	year = lb.lmtime - 6L*30L*24L*60L*60L; /* 6 months ago */
	if (--argc > 0 && *argv[1] == '-') {
		argv++;
		while (*++*argv) switch (**argv) {

		case 'a':
			aflg++;
			continue;

		case 's':
			sflg++;
			statreq++;
			continue;

		case 'd':
			dflg++;
			continue;

		case 'g':
			gflg++;
			continue;

		case 'l':
			lflg++;
			statreq++;
			continue;

		case 'r':
			rflg = -1;
			continue;

		case 't':
			tflg++;
			statreq++;
			continue;

		case 'u':
			uflg++;
			continue;

		case 'c':
			cflg++;
			continue;

		case 'i':
			iflg++;
			continue;

		case 'f':
			fflg++;
			continue;

		default:
			continue;
		}
		argc--;
	}
	if (fflg) {
		aflg++;
		lflg = 0;
		sflg = 0;
		tflg = 0;
		statreq = 0;
	}
	if(lflg) {
		t = "/etc/passwd";
		if(gflg)
			t = "/etc/group";
		pwdf = fopen(t, "r");
	}
	if (argc==0) {
		argc++;
		argv = &dotp - 1;
	}
	for (i=0; i < argc; i++) {
		if ((ep = gstat(*++argv, 1))==NULL)
			continue;
		ep->ln.namep = *argv;
		ep->lflags |= ISARG;
	}
	qsort(firstp, lastp - firstp, sizeof *lastp, compar);
	slastp = lastp;
	for (epp=firstp; epp<slastp; epp++) {
		ep = *epp;
		if (ep->ltype=='d' && dflg==0 || fflg) {
			if (argc>1)
				printf("\n%s:\n", ep->ln.namep);
			lastp = slastp;
			readdir(ep->ln.namep);
			if (fflg==0)
				qsort(slastp,lastp - slastp,sizeof *lastp,compar);
			if (lflg || sflg)
				printf("total %D\n", tblocks);
			for (ep1=slastp; ep1<lastp; ep1++)
				pentry(*ep1);
		} else 
			pentry(ep);
	}
	exit(0);
}

pentry(ap)
struct lbuf *ap;
{
	struct { char dminor, dmajor;};
	register t;
	register struct lbuf *p;
	register char *cp;

	p = ap;
	if (p->lnum == -1)
		return;
	if (iflg)
		printf("%5u ", p->lnum);
	if (sflg)
	printf("%4D ", nblock(p->lsize));
	if (lflg) {
		putchar(p->ltype);
		pmode(p->lflags);
		printf("%2d ", p->lnl);
		t = p->luid;
		if(gflg)
			t = p->lgid;
		if (getname(t, tbuf)==0)
			printf("%-6.6s", tbuf);
		else
			printf("%-6d", t);
		if (p->ltype=='b' || p->ltype=='c')
			printf("%3d,%3d", major((int)p->lsize), minor((int)p->lsize));
		else
			printf("%7ld", p->lsize);
		cp = ctime(&p->lmtime);
		if(p->lmtime < year)
			printf(" %-7.7s %-4.4s ", cp+4, cp+20); else
			printf(" %-12.12s ", cp+4);
	}
	if (p->lflags&ISARG)
		printf("%s\n", p->ln.namep);
	else
		printf("%.14s\n", p->ln.lname);
}

getname(uid, buf)
int uid;
char buf[];
{
	int j, c, n, i;

	if (uid==lastuid)
		return(0);
	if(pwdf == NULL)
		return(-1);
	rewind(pwdf);
	lastuid = -1;
	do {
		i = 0;
		j = 0;
		n = 0;
		while((c=fgetc(pwdf)) != '\n') {
			if (c==EOF)
				return(-1);
			if (c==':') {
				j++;
				c = '0';
			}
			if (j==0)
				buf[i++] = c;
			if (j==2)
				n = n*10 + c - '0';
		}
	} while (n != uid);
	buf[i++] = '\0';
	lastuid = uid;
	return(0);
}

long
nblock(size)
long size;
{
	return((size+1023)>>10);
}

int	m1[] = { 1, S_IREAD>>0, 'r', '-' };
int	m2[] = { 1, S_IWRITE>>0, 'w', '-' };
int	m3[] = { 2, S_ISUID, 's', S_IEXEC>>0, 'x', '-' };
int	m4[] = { 1, S_IREAD>>3, 'r', '-' };
int	m5[] = { 1, S_IWRITE>>3, 'w', '-' };
int	m6[] = { 2, S_ISGID, 's', S_IEXEC>>3, 'x', '-' };
int	m7[] = { 1, S_IREAD>>6, 'r', '-' };
int	m8[] = { 1, S_IWRITE>>6, 'w', '-' };
int	m9[] = { 2, S_ISVTX, 't', S_IEXEC>>6, 'x', '-' };

int	*m[] = { m1, m2, m3, m4, m5, m6, m7, m8, m9};

pmode(aflag)
{
	register int **mp;

	flags = aflag;
	for (mp = &m[0]; mp < &m[sizeof(m)/sizeof(m[0])];)
		select(*mp++);
}

select(pairp)
register int *pairp;
{
	register int n;

	n = *pairp++;
	while (--n>=0 && (flags&*pairp++)==0)
		pairp++;
	putchar(*pairp);
}

char *
makename(dir, file)
char *dir, *file;
{
	static char dfile[100];
	register char *dp, *fp;
	register int i;

	dp = dfile;
	fp = dir;
	while (*fp)
		*dp++ = *fp++;
	*dp++ = '/';
	fp = file;
	for (i=0; i<DIRSIZ; i++)
		*dp++ = *fp++;
	*dp = 0;
	return(dfile);
}

readdir(dir)
char *dir;
{
	static struct direct dentry;
	register int j;
	register struct lbuf *ep;

	if ((dirf = fopen(dir, "r")) == NULL) {
		printf("%s unreadable\n", dir);
		return;
	}
	tblocks = 0;
	for(;;) {
		if (fread((char *)&dentry, sizeof(dentry), 1, dirf) != 1)
			break;
		if (dentry.d_ino==0
		 || aflg==0 && dentry.d_name[0]=='.' &&  (dentry.d_name[1]=='\0'
			|| dentry.d_name[1]=='.' && dentry.d_name[2]=='\0'))
			continue;
		ep = gstat(makename(dir, dentry.d_name), 0);
		if (ep==NULL)
			continue;
		if (ep->lnum != -1)
			ep->lnum = dentry.d_ino;
		for (j=0; j<DIRSIZ; j++)
			ep->ln.lname[j] = dentry.d_name[j];
	}
	fclose(dirf);
}

struct lbuf *
gstat(file, argfl)
char *file;
{
	extern char *malloc();
	struct stat statb;
	register struct lbuf *rep;
	static int nomocore;

	if (nomocore)
		return(NULL);
	rep = (struct lbuf *)malloc(sizeof(struct lbuf));
	if (rep==NULL) {
		fprintf(stderr, "ls: out of memory\n");
		nomocore = 1;
		return(NULL);
	}
	if (lastp >= &flist[NFILES]) {
		static int msg;
		lastp--;
		if (msg==0) {
			fprintf(stderr, "ls: too many files\n");
			msg++;
		}
	}
	*lastp++ = rep;
	rep->lflags = 0;
	rep->lnum = 0;
	rep->ltype = '-';
	if (argfl || statreq) {
		if (stat(file, &statb)<0) {
			printf("%s not found\n", file);
			statb.st_ino = -1;
			statb.st_size = 0;
			statb.st_mode = 0;
			if (argfl) {
				lastp--;
				return(0);
			}
		}
		rep->lnum = statb.st_ino;
		rep->lsize = statb.st_size;
		switch(statb.st_mode&S_IFMT) {

		case S_IFDIR:
			rep->ltype = 'd';
			break;

		case S_IFBLK:
			rep->ltype = 'b';
			rep->lsize = statb.st_rdev;
			break;

		case S_IFCHR:
			rep->ltype = 'c';
			rep->lsize = statb.st_rdev;
			break;
		}
		rep->lflags = statb.st_mode & ~S_IFMT;
		rep->luid = statb.st_uid;
		rep->lgid = statb.st_gid;
		rep->lnl = statb.st_nlink;
		if(uflg)
			rep->lmtime = statb.st_atime;
		else if (cflg)
			rep->lmtime = statb.st_ctime;
		else
			rep->lmtime = statb.st_mtime;
		tblocks += nblock(statb.st_size);
	}
	return(rep);
}

compar(pp1, pp2)
struct lbuf **pp1, **pp2;
{
	register struct lbuf *p1, *p2;

	p1 = *pp1;
	p2 = *pp2;
	if (dflg==0) {
		if (p1->lflags&ISARG && p1->ltype=='d') {
			if (!(p2->lflags&ISARG && p2->ltype=='d'))
				return(1);
		} else {
			if (p2->lflags&ISARG && p2->ltype=='d')
				return(-1);
		}
	}
	if (tflg) {
		if(p2->lmtime == p1->lmtime)
			return(0);
		if(p2->lmtime > p1->lmtime)
			return(rflg);
		return(-rflg);
	}
	return(rflg * strcmp(p1->lflags&ISARG? p1->ln.namep: p1->ln.lname,
				p2->lflags&ISARG? p2->ln.namep: p2->ln.lname));
}
